File: BraintreeApi/src/androidTest/java/com/braintreepayments/api/models/PayPalAccountNonceTest.java
package com.braintreepayments.api.models;

import android.os.Parcel;
import android.support.test.runner.AndroidJUnit4;
import android.test.suitebuilder.annotation.SmallTest;

import org.json.JSONException;
import org.junit.Test;
import org.junit.runner.RunWith;

import static com.braintreepayments.testutils.FixturesHelper.stringFromFixture;
import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class PayPalAccountNonceTest {

    @Test(timeout = 1000)
    public void payPalAccountTypeIsPayPal() {
        assertEquals("PayPal", new PayPalAccountNonce().getTypeLabel());
    }

    @Test(timeout = 1000)
    public void fromJson_parsesResponse() throws JSONException {
        String paypalString = stringFromFixture("payment_methods/paypal_account_response.json");

        PayPalAccountNonce payPalAccountNonce = PayPalAccountNonce.fromJson(paypalString);

        assertEquals("with email paypalaccount@example.com", payPalAccountNonce.getDescription());
        assertEquals("aaaaaa-bbbbbbb-109934023-1", payPalAccountNonce.getNonce());
        assertEquals("paypalaccount@example.com", payPalAccountNonce.getEmail());
        assertEquals("PayPal", payPalAccountNonce.getTypeLabel());
        assertEquals("123 Fake St.", payPalAccountNonce.getBillingAddress().getStreetAddress());
        assertEquals("Apt. 3", payPalAccountNonce.getBillingAddress().getExtendedAddress());
        assertEquals("Oakland", payPalAccountNonce.getBillingAddress().getLocality());
        assertEquals("CA", payPalAccountNonce.getBillingAddress().getRegion());
        assertEquals("94602", payPalAccountNonce.getBillingAddress().getPostalCode());
        assertEquals("US", payPalAccountNonce.getBillingAddress().getCountryCodeAlpha2());
    }

    @Test(timeout = 1000)
    public void getDescription_usesGetEmailIfDescriptionIsPayPalAndEmailIsNotEmpty() {
        PayPalAccountNonce payPalAccountNonce = spy(new PayPalAccountNonce());
        payPalAccountNonce.mDescription = "PayPal";
        when(payPalAccountNonce.getEmail()).thenReturn("test_email_address");

        assertEquals("test_email_address", payPalAccountNonce.getDescription());
    }

    @Test(timeout = 1000)
    public void parcelsCorrectly() throws JSONException {
        String paypalString = stringFromFixture("payment_methods/paypal_account_response.json");
        PayPalAccountNonce payPalAccountNonce = PayPalAccountNonce.fromJson(paypalString);
        Parcel parcel = Parcel.obtain();
        payPalAccountNonce.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        PayPalAccountNonce parceled = PayPalAccountNonce.CREATOR.createFromParcel(parcel);

        assertEquals("with email paypalaccount@example.com", parceled.getDescription());
        assertEquals("aaaaaa-bbbbbbb-109934023-1", parceled.getNonce());
        assertEquals("paypalaccount@example.com", parceled.getEmail());
        assertEquals("PayPal", parceled.getTypeLabel());
        assertEquals("123 Fake St.", parceled.getBillingAddress().getStreetAddress());
        assertEquals("Apt. 3", parceled.getBillingAddress().getExtendedAddress());
        assertEquals("Oakland", parceled.getBillingAddress().getLocality());
        assertEquals("CA", parceled.getBillingAddress().getRegion());
        assertEquals("94602", parceled.getBillingAddress().getPostalCode());
        assertEquals("US", parceled.getBillingAddress().getCountryCodeAlpha2());
    }
}

File: PayPal/src/androidTest/java/com/braintreepayments/api/paypal/PayPalAccountNonceTest.kt
package com.braintreepayments.api.paypal

import android.os.Parcel
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import com.braintreepayments.api.testutils.Fixtures
import kotlinx.parcelize.parcelableCreator
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4ClassRunner::class)
class PayPalAccountNonceTest {

    @Test
    fun fromJSON_withFullResponse_parsesAllFields() {
        val json = JSONObject(Fixtures.PAYMENT_METHODS_PAYPAL_ACCOUNT_RESPONSE)

        val nonce = PayPalAccountNonce.fromJSON(json)

        assertEquals("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee", nonce.string)
        assertFalse(nonce.isDefault)
        assertEquals("paypalaccount@example.com", nonce.email)
        assertEquals("fake-authenticate-url", nonce.authenticateUrl)
        assertEquals("fake-order-id", nonce.paymentId)

        assertNotNull(nonce.creditFinancing)
        assertEquals(18, nonce.creditFinancing?.term)
        assertFalse(nonce.creditFinancing?.isCardAmountImmutable ?: true)
        assertTrue(nonce.creditFinancing?.hasPayerAcceptance ?: false)

        assertNotNull(nonce.billingAddress)
        assertEquals("123 Fake St.", nonce.billingAddress.streetAddress)
        assertEquals("Apt. 3", nonce.billingAddress.extendedAddress)
        assertEquals("Oakland", nonce.billingAddress.locality)
        assertEquals("CA", nonce.billingAddress.region)
        assertEquals("94602", nonce.billingAddress.postalCode)
    }

    @Test
    fun fromJSON_withoutAddresses_returnsEmptyAddresses() {
        val json = JSONObject(Fixtures.PAYMENT_METHODS_PAYPAL_ACCOUNT_RESPONSE_WITHOUT_ADDRESSES)

        val nonce = PayPalAccountNonce.fromJSON(json)

        assertEquals("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee", nonce.string)
        assertEquals("paypalaccount@example.com", nonce.email)
        assertNotNull(nonce.billingAddress)
        assertNotNull(nonce.shippingAddress)
    }

    @Test
    fun fromJSON_withCreditFinancingWithoutFullData_parsesTerm() {
        val json = JSONObject(
            Fixtures.PAYMENT_METHODS_PAYPAL_ACCOUNT_RESPONSE_WITHOUT_CREDIT_FINANCING_DATA
        )

        val nonce = PayPalAccountNonce.fromJSON(json)

        assertEquals("8d4e91d2-4627-10f4-6336-113e331ff9ce", nonce.string)
        assertEquals("paypal@vividseats.com", nonce.email)
        assertNotNull(nonce.creditFinancing)
        assertEquals(18, nonce.creditFinancing?.term)
        assertNotNull(nonce.creditFinancing?.monthlyPayment)
        assertNull(nonce.creditFinancing?.monthlyPayment?.currency)
        assertNull(nonce.creditFinancing?.monthlyPayment?.value)

        assertEquals("Vivid", nonce.firstName)
        assertEquals("Tester", nonce.lastName)
        assertEquals("408-242-6641", nonce.phone)
        assertEquals("NSXGUJGV6WMGW", nonce.payerId)

        assertNotNull(nonce.billingAddress)
        assertEquals("1 Main St", nonce.billingAddress.streetAddress)
        assertEquals("San Jose", nonce.billingAddress.locality)
    }

    @Test
    fun parcels_correctly() {
        val json = JSONObject(Fixtures.PAYMENT_METHODS_PAYPAL_ACCOUNT_RESPONSE)
        val original = PayPalAccountNonce.fromJSON(json)

        val parcel = Parcel.obtain()
        original.writeToParcel(parcel, 0)
        parcel.setDataPosition(0)

        val restored = parcelableCreator<PayPalAccountNonce>().createFromParcel(parcel)
        parcel.recycle()

        assertEquals(original.string, restored.string)
        assertEquals(original.isDefault, restored.isDefault)
        assertEquals(original.email, restored.email)
        assertEquals(original.authenticateUrl, restored.authenticateUrl)
        assertEquals(original.paymentId, restored.paymentId)
        assertEquals(original.billingAddress.streetAddress, restored.billingAddress.streetAddress)
        assertEquals(original.billingAddress.locality, restored.billingAddress.locality)
        assertEquals(original.creditFinancing?.term, restored.creditFinancing?.term)
        assertEquals(
            original.creditFinancing?.isCardAmountImmutable,
            restored.creditFinancing?.isCardAmountImmutable
        )
    }
}
